### ISSUE TITLE
libkodein-leveldb-jni.dll: Can't find dependent libraries

### ISSUE THREAD

**Issue by SimonBurfield** [Kodein-Framework/Kodein-DB#28]
We are using Kodein-DB in our Android app and would like to write a console app (Kotlin) to read the DB on Windows in case we need to check what is in it from a support angle.

When we run the following code we receive the following error 
```
val dbDirectory = "C:\\db\\Test"
    val db = DB.open(dbDirectory,
        KotlinxSerializer {
            +ReferenceDataResult.ReferenceDataResultSerializer
        },
        OpenPolicy.OpenOrCreate
    )
    println("Opened DB!") 
```

```
Exception in thread "main" java.lang.UnsatisfiedLinkError: C:\Users\simon.burfield\AppData\Local\Kodein-DB\0.6.0-beta\libkodein-leveldb-jni.dll: Can't find dependent libraries
	at java.lang.ClassLoader$NativeLibrary.load(Native Method)
	at java.lang.ClassLoader.loadLibrary0(ClassLoader.java:1946)
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1828)
	at java.lang.Runtime.load0(Runtime.java:810)
	at java.lang.System.load(System.java:1088)
	at org.kodein.db.leveldb.jvm.AbstractLevelDBJvmLoader.load(AbstractLevelDBJvmLoader.kt:77)
	at org.kodein.db.leveldb.jvm.LevelDBJvm.<clinit>(LevelDBJvm.kt:28)
	at org.kodein.db.impl.data.DataDBJvm.getLdbFactory(DataDBJvm.kt:10)
	at org.kodein.db.impl.data.AbstractDataDBFactory.open(AbstractDataDBFactory.kt:18)
	at org.kodein.db.impl.data.AbstractDataDBFactory.open(AbstractDataDBFactory.kt:11)
	at org.kodein.db.impl.model.AbstractModelDBFactory.open(AbstractModelDBFactory.kt:29)
	at org.kodein.db.impl.model.AbstractModelDBFactory.open(AbstractModelDBFactory.kt:9)
	at org.kodein.db.impl.AbstractDBFactory.open(AbstractDBFactory.kt:25)
	at org.kodein.db.impl.AbstractDBFactory.open(AbstractDBFactory.kt:10)
	at org.kodein.db.impl.DefaultKt.open(default.kt:7)
	at MainKt.main(main.kt:13)

```

**Comment by SalomonBrys** [comment]
Hi @SimonBurfield,
which version of Kodein-DB are you using ?
We fixed a similar issue some time ago, so are you sure you are using the latest version ?

**Comment by nordiauwu** [comment]
> Hi @SimonBurfield,
> which version of Kodein-DB are you using ?
> We fixed a similar issue some time ago, so are you sure you are using the latest version ?

They are using 0.6.0-beta (you can see it in exception message), anyway I just tested with the latest 0.6.1 and still getting the exact same error

**Comment by SimonBurfield** [comment]
Yes, I used 0.6.0 and could not get it to work on Windows 10. 

Would be great to fix this

**Comment by zTrap** [comment]
This error can depends on OS. I have same issue with any library version on my Win Pro x64

**Comment by SimonBurfield** [comment]
So can we fix it?

**Comment by romainbsl** [comment]
This seems like a build issue. We will try to reproduce the issue to fix it in a future version.

**Comment by zTrap** [comment]
Do you need some specific system information?

**Comment by romainbsl** [comment]
A reproducer might be helpful, but my guess is that creating a simple Windows console app if a good start to investigate the issue.

**Comment by SimonBurfield** [comment]
On windows this fails on the Open

```
fun main(args: Array<String>) {
    println("Hello World!")

    val dbDirectory = "C:\\db\\Test"
    val db = DB.open(dbDirectory,
        KotlinxSerializer {
            +ReferenceDataResult.ReferenceDataResultSerializer
        },
        OpenPolicy.OpenOrCreate
    )
    println("Opened DB!")

}

data class ReferenceDataResult(
        @Id
        val id: String,

        val expiresOn: Long,

        val referenceDataType: String,

        val value: JsonObject
) {
    object ReferenceDataResultSerializer : Serializer<ReferenceDataResult> {

        override fun deserialize(
                type: KClass<out ReferenceDataResult>,
                transientId: ReadMemory,
                input: ReadBuffer,
                vararg options: Options.Read
        ): ReferenceDataResult {
            val id = input.readString()
            val expiresOn = input.readLong()
            val referenceDataType = input.readString()
            val value = Json.decodeFromString<JsonObject>(input.readString())//Gson().fromJson(input.readString(), JsonObject::class.java).asJsonObject

            return ReferenceDataResult(id, expiresOn, referenceDataType, value)
        }

        override fun serialize(model: ReferenceDataResult, output: Writeable, vararg options: Options.Write) {
            output.putString(model.id)
            output.putLong(model.expiresOn)
            output.putString(model.referenceDataType)
            output.putString(model.value.toString())
        }
    }
}
```

**Comment by romainbsl** [comment]
Here is our reproducer https://github.com/romainbsl/kodein-db-issue-28-reproducer, we will start from there. 
We certainly have to investigate our build process why the dll is not package with every other artifacts.
We will keep you posted.

**Comment by SalomonBrys** [comment]
The DLL **is** packaged.
The problem is missing dependencies of that DLL.

Until Kodein 0.6 (included), the Kodein-DB DLL (containing LevelDB & JNI bindings) is compiled with MinGW64 for the `x86_64-pc-windows-gnu` target. This was the "easiest" way, because Kotlin/Native is using MinGW64 so we could compile for Kotlin/Native & Windows-JVM using the same compilers.
However, compiling for this target apparently adds some MinGW specific dependencies, which we did not see since we tested on our development machines, which obviously already contained the required dependencies.

**To have Kodein 0.6 working on Windows, you need to have [these 3 DLLs](https://github.com/Kodein-Framework/Kodein-DB/files/6353158/Kodein-DB-DLL-Dependencies.zip) either in the working directory of the app when it launches, or in any directory of the PATH environment variable.**

This is obviously not ideal.
In Kodein 0.7, the Kodein-DB DLL is compiled with Visual Studio, which does not add any dependency that's not already part of the Windows OS. I've tested it on a blank virtual Windows to confirm.
You can already test it with the [latest](https://oss.sonatype.org/content/repositories/snapshots/org/kodein/db/kodein-db/0.7.0-beta-master-134-SNAPSHOT/) [snapshot](https://oss.sonatype.org/content/repositories/snapshots/org/kodein/db/kodein-db/0.7.0-beta-master-134-SNAPSHOT/).

To recap:
- This issue is fixed in Kodein 0.7
- You need to add [these 3 DLLs](https://github.com/Kodein-Framework/Kodein-DB/files/6353158/Kodein-DB-DLL-Dependencies.zip) along your app when using Kodein 0.6 and lower

I'll close this issue when Kodein 0.7 is released.

**Comment by SalomonBrys** [comment]
Kodein 0.8.0-beta is released and fixes this issue.
